packages/vscode-integration/src/common/glsp-vscode-connector.ts

```diff
@@ -256,13 +239,59 @@ export class GlspVscodeConnector<D extends vscode.CustomDocument = vscode.Custom
      *
      * @param clientId Id of client.
      * @param action Action to send.
+     *
+     * @deprecated Use  {@link dispatchAction} instead.
      */
     protected sendActionToClient(clientId: string, action: Action): void {
-        this.sendMessageToClient(clientId, {
-            clientId: clientId,
-            action: action,
-            __localDispatch: true
-        });
+        this.dispatchAction(action, clientId);
+    }
+
+    /**
+     * Send an action to the client/panel that is currently focused. If no registered
+     * panel is focused, the message will not be sent.
+     *
+     * @param action The action to send to the active client.
+     * @deprecated Use {@link dispatchAction} instead.
+     */
+    public sendActionToActiveClient(action: Action): void {
+        this.dispatchAction(action);
+    }
+
+    /**
+     * Dispatches an action associated with the given client id. If no id is provided,
+     * the action will be dispatched associated with the client of the active webview panel.
+     * If no client id is passed and no registered panel is focused, the action will not be dispatched.
+     * Dispatching an action will send the action to the client and/or server if
+     * they can handle the action.
+     * @param action The action to dispatch.
+     * @param clientId The id of the client/session associated with the action.
+     */
+    dispatchAction(action: Action, clientId?: string): void {
+        const client = clientId ? this.clientMap.get(clientId) : this.getActiveClient();
+        if (!client) {
+            return;
```

Comment:
I think it would be a great idea to log that we cannot find the client with the given id to send the given action, just so in case something goes wrong it might be easier to find the source of the problem.

Reply:
Added with the latest commit